### DISCUSSION TITLE
How to install and/or use another version of ParaView

### DISCUSSION THREAD

**Discussion by wyldckat** [blueCFD/Core#200] [Q&A] [answered]
## Reason

There are two main reasons for wanting to use another version of ParaView:

 1. If ParaView crashes right after launching it, it is very likely that either the GPU/Graphics card in your machine doesn't support OpenGL 2, or that the drivers currently installed for that GPU do not properly support OpenGL 2.

    * In this scenario, it's best to install ParaView 4.4.0, which was the last pre-built version for Windows that provided OpenGL 1.x support.

 2. If the version of ParaView provided with blueCFD-Core is not the latest version.

**Comment by wyldckat** [comment, marked as answer]
There are at least 2 possible solutions:

 1. [Replace the installation of ParaView in blueCFD-Core](#solution-1)
 2. [Use an existing installation of ParaView](#solution-2)

**Note:** Both solutions require you to have installed ParaView through the blueCFD-Core installer, so that the environment is automatically loaded into the blueCFD-Core terminal and allow ParaView to be launched from the terminal's command line.


### <a name="solution-1">Solution 1 - Replace the installation of ParaView in blueCFD-Core</a>

 1. Start the blueCFD-Core terminal.

 2. Execute the following command in the terminal:

    ```
    cd ~/AddOns/
    ```

 3. Start Windows Explorer, because the next steps have to be done interactively, since administrative permission may be necessary:

    ```
    explorer .
    ```

    Notice that the command is: `explorer` _space_ _dot_

 4. Windows Explorer should open up and show the contents of the `AddOns` folder.

 5. Now either rename or delete the contents of the folder `ParaView`. Don't forget to create a new `ParaView` folder, so that it can be used with the ParaView installer.

 6. Now you can inside the `ParaView` folder and copy the full path from the address bar and use it during the installation of ParaView. Below is shown the address bar from Windows 10; make sure to click on the right side of the path so that it goes into editing mode that is shown below:

    ![Example of where to copy the path from](http://bluecfd.github.io/Core/images/ParaView_Path_to_copy.png)

 7. Run the ParaView installer for the version you want to use and be certain that the correct path is used, as shown in the image below:

    ![Example of where to change the installation path](http://bluecfd.github.io/Core/images/Install_Path_For_ParaView.png)

 8. Once ParaView is installed, it should be ready to be used from the blueCFD-Core terminal.


### <a name="solution-2">Solution 2 - Use an existing installation of ParaView</a>

**Note**: The downside to using this solution is that the file extension `.foam` is not automatically register in Windows to open with this version of ParaView.

 1. Using as an example, if ParaView 4.4 is installed in the folder `C:\Program Files\ParaView 4.4.0`.

 2. Start the blueCFD-Core terminal.

 3. Execute the following command in the terminal:

    ```
    nano ~/.bashrc.d/paraview.sh
    ```

 4. It should have opened the file in a window of Notepad2.

 5. Now it depends on whether you install ParaView that comes with blueCFD-Core or now:

    * If you installed ParaView that comes with blueCFD-Core, then go to line 42 and change this line:

        ```
        export AddOns_ParaView_DIR=$HOME/AddOns/ParaView
        ```

        To this line:

        ```
        export AddOns_ParaView_DIR="/c/Program Files/ParaView 4.4.0"
        ```

    * If you did not install ParaView that comes with blueCFD-Core, then that means that you are looking at an empty file, in which case you can simply type in the following two lines:

        ```
        export AddOns_ParaView_DIR="/c/Program Files/ParaView 4.4.0"
        export PATH=$PATH:$AddOns_ParaView_DIR/bin
        ```

 8. Save the file (menu `File --> Save`) and close the Notepad2 window.

 9. Use the key combination `Alt+F2` to start a new terminal. If there are no error messages, then the older terminal can now be closed.

 10. From now on, whenever the commands `paraview` or `paraFoam` are executed, ParaView 4.4 should always open.

**Comment by salicto** [comment]
I'm currently encountering an issue with the 'paraview.sh' file in my setup for Blue CFD. It appears that the file is empty, and I'm unable to find the necessary instructions to specify the path for Paraview.

I already have a version of Paraview 5.11.1 installed, and I would like to link the application with BlueCFD.


Thank you in advance for your assistance.

Best regards,

**Reply by wyldckat** [reply]
@salicto A few quick questions:

1. Which version of blueCFD-Core are you using?
2. In which command line interface did you type the commands for the solution 2?

**Reply by salicto** [reply]
Hi @wyldckat,

1. I use the 2020-1
2. I type the commands in the bluecfd terminal

**Reply by wyldckat** [reply]
Sorry, I should have asked in the previous post... please show an image of the terminal and the exact command used, just in case. One small detail, even a dot in the wrong place, is enough for this to not work.

In addition to this, please run this command in the same terminal:
```
ll ~/.bashrc.d/
```

**Reply by salicto** [reply]
When I type your command, I get this :
![Terminal_blueCFD](https://github.com/blueCFD/Core/assets/57439956/5c322a44-a1b6-4af7-8df3-cf83e2a2d84d)
 When I type `nano ~/.bashrc.d/`, it opens a window with only a notepad2.sh script. And if I write `nano ~/.bashrc.d/paraview.sh`, it creates an empty paraview script. I tried to fill it with the line: `export AddOns_ParaView_DIR="/c/Program Files/ParaView 5.11.1".`

**Reply by wyldckat** [reply]
OK, I've finally collected the dots. Given that you hadn't installed the ParaView version that comes with blueCFD-Core, that means that this `paraview.sh` file is non-existing.

The line you are missing is for setting the PATH environment variable, namely add the following line after that one:
```
export PATH=$PATH:$AddOns_ParaView_DIR/bin
```

I will update the documentation above as well.

edit: I've updated the instructions above.

**Reply by salicto** [reply]
Okay, thank you. I made the requested modification. I tried it on a simple test case. I type "BlockMesh" and then "paraFoam &" and I get an error. The program successfully creates the file "TutorialCase.foam," but it is still not possible to open ParaView. The issue lies with the paraFoam command file. What should I modify?
[parafoam.bat.txt](https://github.com/blueCFD/Core/files/11845770/parafoam.bat.txt)
![Parafoam_error](https://github.com/blueCFD/Core/assets/57439956/ff6ec781-ab21-44e3-af48-cfc9d3c03eb6)

**Reply by wyldckat** [reply]
Quick questions:
1. Did you start a new terminal window, after making the changes to that file, as explained in the the instructions for solution 2 above?

2. What do the following commands give you?
```
cat ~/.bashrc.d/paraview.sh

which paraview
where paraview

ls "/c/Program Files/ParaView 5.11.1/bin"/*.exe
```

**Comment by salicto** [comment]
Yes I start a new terminal. Ok there's the terminal output
![Terminal](https://github.com/blueCFD/Core/assets/57439956/1ae34104-2b79-4ee5-a86c-f8b2d08151c3)

**Reply by wyldckat** [reply]
Please try again running in the tutorial folder the same command as before:
```
paraFoam &
```

**Reply by salicto** [reply]
It's still not working. It's not a big deal though, the IT department of the company allows me to use a virtual machine after all. I will install OpenFOAM directly on Ubuntu. Thank you again for your time.

**Reply by wyldckat** [reply]
If you still want to try and fixing this, I had to re-read your previous post about the `paraFoam` script not working... since you had sent the other file, I forgot to check the actual one (which doesn't have any extension in its name).

My suspicion is that the problem is that the path for where ParaView was installed, has spaces in it, but the paraFoam script is unable to handle those spaces. The trick should be to replace the path with the shorthand version of it, assuming Windows is still supporting this feature in your installation. If you run this command in the terminal:
```
cygpath -m -s "/c/Program Files/ParaView 5.11.1"
```

it should give something like this:
```
C:/PROGRA~1/PARAVI~1.1
```

With this you can then modify the respective line (the tilde `~` characters should not have any spaces):
```
export AddOns_ParaView_DIR="/c/PROGRA~1/PARAVI~1.1"
```

then this should work, after starting a new terminal.

If it still doesn't work... you can try running these commands, instead of running `paraFoam`:
```
touch case.foam
paraview
```
then ParaView should open, but you'll have to open the `case.foam` file manually via ParaView.
If this works, then it's possible to simplify this further and make it so that this is done automatically with a single command.
